### PR TITLE
feat(numeral-date): add date parts internal validation - FE-3184

### PR DESCRIPTION
**Pull request by mkrds**
### Proposed behaviour
This PR adds internal validation to NumeralDate component which has following rules:
Day: value within range 1-31,
Month: value within range 1-12,
Year: value within range 1800-2200

This validation is triggered when you leave one of the date partial inputs - i.e day field is blurred with incorrect value -> validation for day field is triggered and displayed.

If external `error` is provided the internal one will be overriden.

### Current behaviour
Currently only external validation is available.

### Checklist

- [x] Commits follow our style guide
- [x] Screenshots are included in the PR if useful
- [x] All themes are supported if required
- [x] Unit tests added or updated if required
- [x] Cypress automation tests added or updated if required
- [x] Storybook added or updated if required
- [x] Typescript `d.ts` file added or updated if required
- [x] Carbon implementation and Design System documentation are congruent

### Testing instructions
Any story where NumeralDate has no `error` prop set already.

### CODE REVIEW

**Review by sjohnson-sage — APPROVED** [review]

**Review by edleeks87 — APPROVED** [review]

**Review by Alec-Walker-Sage — COMMENTED** [review]
I can't decide whether cypress tests are needed because seems to cover the functionality in the unit tests, need advice from @DlgSHi or @KatarzynaQA :)

File: src/__experimental__/components/numeral-date/numeral-date.component.js [hunk 1]
```diff
@@ -30,11 +55,17 @@ const NumeralDate = ({
   fieldHelp,
   adaptiveLabelBreakpoint,
   required,
+  enableInternalError,
+  enableInternalWarning,
```
[the remark this answers is not in the document]
Comment (mkrds) [reply src/__experimental__/components/numeral-date/numeral-date.component.js:59]: Yeah, I've discussed it with @harpalsingh

File: cypress/support/step-definitions/numeral-date-steps.js [hunk 2]
```diff
@@ -31,3 +35,8 @@ Then('Date format in {string} input is set to {word}', (position, text) => {
   numeralDateInputByPosition(positionOfElement(position)).should('have.attr', 'placeholder', text);
   numeralDateInputByPosition(positionOfElement(position)).should('have.attr', 'placeholder', text);
 });
+
+Then('Warning validation is visible', () => {
+  icon().should('have.attr', 'type', 'warning');
+  icon().should('be.visible');
```
Comment (DlgSHi) [inline cypress/support/step-definitions/numeral-date-steps.js:41]: there is no need to call icon() one more time, we can simplify this and add:

``` 
icon().should('have.attr', 'type', 'warning').and('be.visible');
```

File: cypress/locators/numeralDate/locators.js [hunk 3]
```diff
@@ -1 +1,2 @@
 export const NUMERAL_DATE_COMPONENT = '[data-component="numeral-date"]';
+export const ICON = 'span[data-component="icon"]';
```
Comment (DlgSHi) [inline cypress/locators/numeralDate/locators.js:2]: we have already defined `ICON` in `common-steps.js`

**Review by DlgSHi — APPROVED** [review]

**Review by nicktitchmarsh — APPROVED** [review]

**Review by ianforster-sage — APPROVED** [review]

File: cypress/support/step-definitions/numeral-date-steps.js [hunk 4]
```diff
@@ -11,6 +11,10 @@ Then('I type numeral characters {string} in {string} inputs', (text, position) =
   numeralDateInputByPosition(positionOfElement(position)).parent().should('have.css', 'outline-color', 'rgb(255, 181, 0)');
 });
 
+When('I focus on input {string} and blur', (position) => {
+  numeralDateInputByPosition(positionOfElement(position)).clear().blur();
```
Comment (DlgSHi) [inline cypress/support/step-definitions/numeral-date-steps.js:15]: we need to `clear()` or to `focus()` ?

File: cypress/support/step-definitions/numeral-date-steps.js [hunk 5]
```diff
@@ -40,3 +44,7 @@ Then('Date format in {string} input is set to {word}', (position, text) => {
 Then('Warning validation is visible', () => {
   icon().should('have.attr', 'type', 'warning').and('be.visible');
 });
+
+Then('Warning should not be visible', () => {
+  icon().should('not.be.visible');
```
Comment (DlgSHi) [inline cypress/support/step-definitions/numeral-date-steps.js:49]: we need to precise which exactly icon should not be visible (warning in our case)

File: cypress/features/regression/designSystem/numeralDate.feature [hunk 6]
```diff
@@ -25,4 +25,30 @@ Feature: Design System Numeral Date component
...
+      | position | string |
+      | first    | 0      |
+      | first    | 32     |
+      | second   | 00     |
+      | second   | 13     |
+      | third    | 0000   |
+      | third    | 1799   |
+      | third    | 2201   |
+
+  @posative
```
Comment (DlgSHi) [inline cypress/features/regression/designSystem/numeralDate.feature:45]:
```suggestion
  @positive
```

**Review by DlgSHi — DISMISSED** [review]

**Review by Alec-Walker-Sage — DISMISSED** [review]

**Review by Alec-Walker-Sage — APPROVED** [review]

**Review by DlgSHi — APPROVED** [review]

**Review by sjohnson-sage — APPROVED** [review]

**Pull request merged**